File: 1-Two Sum/TwoSum.kt
fun main() {
    val target = 9
    val nums = intArrayOf(2, 7, 11, 15)
    val result = Solution().twoSum(nums, target)

    println(result.toList())
}

class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        // First attempt Mar 15, 2024 05:56
        // Runtime 200ms, Memory 40.19MB
        //
//        val hashTemp = hashMapOf<Int, Int>()
//        var temp = intArrayOf()
//
//        for ((index, value) in nums.withIndex()) {
//            val diffs = target - value
//
//            if (hashTemp.contains(diffs)) {
//                temp += hashTemp[diffs] ?: 0
//                temp += index
//            } else {
//                hashTemp[value] = index
//            }
//        }
//
//        return temp
        //
        // Second attempt Apr 16, 2024 21:44
        // Runtime 188ms, Memory 38.08MB
        //
//        val hashTemp = hashMapOf<Int, Int>()
//        var temp = intArrayOf()
//
//        for (i in nums.indices) {
//            val diffs = target - nums[i]
//
//            if (hashTemp.keys.contains(diffs)) {
//                temp = intArrayOf(hashTemp[diffs] ?: 0, i)
//            }
//
//            hashTemp[nums[i]] = i
//        }
//
//        return temp
        //
        var indexes = intArrayOf()
        for (i in nums.indices) {
            for (j in i + 1..<nums.size) {
                if (nums[i] + nums[j] == target) {
                    indexes = intArrayOf(i, j)
                }
            }
        }

        return indexes
    }
}

File: 1-Two Sum/two_sum.go
package main

import "fmt"

func main() {
	target := 6
	nums := []int{3, 2, 4}

	fmt.Println("Result: ", twoSum(nums, target))
}

func twoSum(nums []int, target int) []int {
	// First attempt Apr 16, 2024 20:26
	// Runtime 6ms, Memory 4.17MB
	//
	mapResult := make(map[int]int)

	for index, num := range nums {
		if _, ok := mapResult[num]; ok {
			return []int{mapResult[num], index}
		}

		diff := target - num
		mapResult[diff] = index
	}

	return []int{}
	//
	// Second attempt Apr 16, 2024 20:46
	// Runtime 8ms, Memory 5.49MB
	//
	//	mapDiffs := make(map[int]int)
	//	var indexes []int
	//
	//	for index, num := range nums {
	//		diff := target - num
	//
	//		if _, ok := mapDiffs[diff]; ok {
	//			indexes = []int{mapDiffs[diff], index}
	//		}
	//
	//			mapDiffs[num] = index
	//		}
	//
	//		return indexes
	//
	// Third attempt Apr 16, 2024 21:00
	// Runtime 7ms, Memory 5.52MB
	//	mapDiffs := make(map[int]int)
	//	var indexes []int
	//
	//	for index, num := range nums {
	//		diff := target - num
	//
	//		if n, ok := mapDiffs[diff]; ok {
	//			indexes = []int{n, index}
	//		}
	//
	//		mapDiffs[num] = index
	//	}
	//
	//	return indexes
}
